test/test.highlevelConsumer.js

```diff
@@ -0,0 +1,222 @@
+    setTimeout(cb, 0);
+  };
+  this.deletePartitionOwnership = function (groupId, topic, partition, cb) {
+    setTimeout(cb, 0);
+  };
+  this.addPartitionOwnership = function (id, groupId, topic, partition, cb) {
+    setTimeout(cb, 0);
+  };
+  this.registerConsumer = function (groupId, id, payloads, cb) {
+    setTimeout(cb, 0);
+  };
+  this.listConsumers = function (groupId) {};
+  this.getConsumersPerTopic = function (groupId, cb) {
+    var consumerTopicMap = {};
+    consumerTopicMap[groupId] = [ 'fake-topic' ];
+    var topicConsumerMap = { 'fake-topic': [ groupId ] };
+    var topicPartitionMap = { 'fake-topic': [ '0', '1', '2' ] };
+    var map = {
+      consumerTopicMap: consumerTopicMap,
+      topicConsumerMap: topicConsumerMap,
+      topicPartitionMap: topicPartitionMap
+    };
+
+    setTimeout(
+      cb.bind(null, null, map),
+      0
+    );
+  };
+}
+util.inherits(FakeZookeeper, EventEmitter);
+
+function FakeClient () {
+  EventEmitter.call(this);
+
+  this.zk = new FakeZookeeper();
+
+  this.topicExists = function (topics, cb) {
+    setTimeout(cb, 0);
+  };
+  this.refreshMetadata = function (topicNames, cb) {
+    setTimeout(cb, 0);
+  };
+  this.sendOffsetCommitRequest = function (groupId, commits, cb) {
+    setTimeout(cb, 0);
+  };
+  this.sendFetchRequest = function (consumer, payloads, fetchMaxWaitMs, fetchMinBytes, maxTickMessages) {};
+  this.sendOffsetFetchRequest = function (groupId, payloads, cb) {
+    setTimeout(cb, 0);
+  };
+  this.sendOffsetRequest = function (payloads, cb) {
+    setTimeout(cb, 0);
+  };
+  this.addTopics = function (topics, cb) {
+    setTimeout(cb, 0);
+  };
+  this.removeTopicMetadata = function (topics, cb) {
+    setTimeout(cb, 0);
+  };
+  this.close = function (cb) {
+    setTimeout(cb, 0);
+  };
+}
+util.inherits(FakeClient, EventEmitter);
+
+describe('HighLevelConsumer', function () {
+  describe('rebalance', function () {
+    var client,
+      highLevelConsumer,
+      sandbox;
+
+    beforeEach(function () {
+      client = new FakeClient();
+
+      highLevelConsumer = new HighLevelConsumer(
+        client,
+        [ {topic: 'fake-topic'} ]
+      );
+
+      clearTimeout(highLevelConsumer.checkPartitionOwnershipInterval);
+
+      sandbox = sinon.sandbox.create();
+    });
+
+    afterEach(function () {
+      highLevelConsumer.close(function () {});
+      sandbox.restore();
+      client = null;
+      highLevelConsumer = null;
+    });
+
+    it('should emit rebalanced event and clear rebalancing flag only after offsets are updated', function (done) {
+      client.emit('ready');
+
+      highLevelConsumer.on('registered', function () {
+        var sendOffsetFetchRequestCalled = false;
+        var refreshMetadataCalled = false;
+
+        // verify rebalancing is false until rebalance finishes
+        var refreshMetadata = client.refreshMetadata;
+        client.refreshMetadata = function () {
```

Comment:
This can be shortened by using a stub. And later you can `sinon.assert` that this stub was called.

```js
        sandbox.stub(client, 'sendOffsetFetchRequest', function (groupId, payloads, cb) {
          highLevelConsumer.rebalancing.should.be.true;
          setImmediate(cb);
        });
```